### PR TITLE
[CPP Graph] Falcon MHA support

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: feature
API not changed

## Description
Add MHA fusion support for Falcon, including:
- Support GQA and MQA
- Replace is_causal by attn_flags for future flags (particularly prepared for MPT's alibi)
- Forward the `--memory-f16/f32/auto` tags to the python script
- Add `NE_OP_DEBUG` for debugging experience

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
<details>
<summary>Test commands:</summary>

```bash
set -ex
#main branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-auto -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"

#this branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-auto -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-f16 -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_falcon && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-f16 -p "$LUOYU_PROMPT"
```
</details>

In conclusion, the model can generate reasonable text with MHA fusion enabled, and it preserves the identical behavior when disable the fusion (`--memory-f16`). The fusion brings ~80% (33833.79ms => 19023.91ms) for first token and ~20% (64.66ms => 53.95ms) performance for next token inference.

### Main branch 
- <details>
  <summary>
  <code>numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-auto -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But most of all, she wanted to learn. And so she did.
  This little girl grew up, and became a writer (and sometimes a photographer) and she's still learning.
  She learned that she must live each moment and enjoy it, no matter how hard the road may be.
  She
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  81.301 ms
  perf_total_per_op_us[                     MUL] =  26.273 ms
  perf_total_per_op_us[                  REPEAT] = 297.837 ms
  perf_total_per_op_us[                    GELU] = 408.667 ms
  perf_total_per_op_us[                    NORM] =  29.996 ms
  perf_total_per_op_us[                 MUL_MAT] = 29184.555 ms
  perf_total_per_op_us[                   SCALE] = 2971.997 ms
  perf_total_per_op_us[                     CPY] =  44.583 ms
  perf_total_per_op_us[                    VIEW] =   0.896 ms
  perf_total_per_op_us[                 PERMUTE] =   0.502 ms
  perf_total_per_op_us[                GET_ROWS] =   9.409 ms
  perf_total_per_op_us[           DIAG_MASK_INF] = 183.420 ms
  perf_total_per_op_us[                SOFT_MAX] = 530.065 ms
  perf_total_per_op_us[                    ROPE] =  59.033 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.644 ms
  perf_total_per_op_us[                     MUL] =   0.219 ms
  perf_total_per_op_us[                    GELU] =  12.005 ms
  perf_total_per_op_us[                    NORM] =   1.533 ms
  perf_total_per_op_us[                 MUL_MAT] =   9.529 ms
  perf_total_per_op_us[                   SCALE] =   1.491 ms
  perf_total_per_op_us[                     CPY] =   4.314 ms
  perf_total_per_op_us[                    VIEW] =   0.754 ms
  perf_total_per_op_us[                 PERMUTE] =   0.459 ms
  perf_total_per_op_us[                GET_ROWS] =   0.027 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.122 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.853 ms
  perf_total_per_op_us[                    ROPE] =   1.176 ms
  perf_total_per_op_us[           INNER PRODUCT] =  30.112 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.543 ms
  perf_total_per_op_us[                     MUL] =   0.201 ms
  perf_total_per_op_us[                    GELU] =  11.318 ms
  perf_total_per_op_us[                    NORM] =   1.264 ms
  perf_total_per_op_us[                 MUL_MAT] =   8.251 ms
  perf_total_per_op_us[                   SCALE] =   1.447 ms
  perf_total_per_op_us[                     CPY] =   3.837 ms
  perf_total_per_op_us[                    VIEW] =   0.810 ms
  perf_total_per_op_us[                 PERMUTE] =   0.415 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.119 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.672 ms
  perf_total_per_op_us[                    ROPE] =   1.059 ms
  perf_total_per_op_us[           INNER PRODUCT] =  29.651 ms
  ========================================
  model_print_timings:        load time = 34974.57 ms
  model_print_timings:      sample time =     2.66 ms /     3 runs   (    0.89 ms per token)
  model_print_timings: prompt eval time = 33833.78 ms /  2008 tokens (   16.85 ms per token)
  model_print_timings:        eval time =   132.93 ms /     2 runs   (   66.47 ms per token)
  model_print_timings:       total time = 35165.28 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 33833.79ms
  prediction   1, time: 68.28ms
  prediction   2, time: 64.66ms
  ```
  </details>

### This branch
- <details>
  <summary>
  <code>numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-auto -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But this little girl lived with her parents and her brother. Her parents had a lot of rules for her. For example she couldn't go out alone after 7pm at night.
  The little girl was always very good to her parents, always listened to them. So she tried her best to obey all the
  ```
  </details>
- <details>
  <summary>
  <code>numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 64 --memory-f16 -p "$GIRL_PROMPT"</code>
  </summary>

  ```
  Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But most of all, she wanted to learn. And so she did.
  This little girl grew up, and became a writer (and sometimes a photographer) and she's still learning.
  She learned that she must live each moment and enjoy it, no matter how hard the road may be.
  She
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  78.572 ms
  perf_total_per_op_us[                     MUL] =  28.676 ms
  perf_total_per_op_us[                  REPEAT] = 282.013 ms
  perf_total_per_op_us[                    GELU] = 478.248 ms
  perf_total_per_op_us[                    NORM] =  27.356 ms
  perf_total_per_op_us[                 MUL_MAT] = 19122.167 ms
  perf_total_per_op_us[                     CPY] =   9.288 ms
  perf_total_per_op_us[                    VIEW] =   0.904 ms
  perf_total_per_op_us[                 PERMUTE] =   0.129 ms
  perf_total_per_op_us[                GET_ROWS] =   9.050 ms
  perf_total_per_op_us[                    ROPE] =  51.539 ms
  perf_total_per_op_us[              FLASH_ATTN] = 262.149 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  14.172 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.591 ms
  perf_total_per_op_us[                     MUL] =   0.212 ms
  perf_total_per_op_us[                    GELU] =  11.214 ms
  perf_total_per_op_us[                    NORM] =   1.203 ms
  perf_total_per_op_us[                     CPY] =   0.897 ms
  perf_total_per_op_us[                    VIEW] =   0.881 ms
  perf_total_per_op_us[                 PERMUTE] =   0.115 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[                    ROPE] =   1.029 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.219 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.551 ms
  perf_total_per_op_us[           INNER PRODUCT] =  31.187 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.474 ms
  perf_total_per_op_us[                     MUL] =   0.176 ms
  perf_total_per_op_us[                    GELU] =  11.143 ms
  perf_total_per_op_us[                    NORM] =   1.238 ms
  perf_total_per_op_us[                     CPY] =   0.884 ms
  perf_total_per_op_us[                    VIEW] =   0.871 ms
  perf_total_per_op_us[                 PERMUTE] =   0.107 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[                    ROPE] =   1.030 ms
  perf_total_per_op_us[              FLASH_ATTN] =   3.791 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.371 ms
  perf_total_per_op_us[           INNER PRODUCT] =  30.322 ms
  ========================================
  model_print_timings:        load time = 21563.13 ms
  model_print_timings:      sample time =     3.23 ms /     3 runs   (    1.08 ms per token)
  model_print_timings: prompt eval time = 20368.36 ms /  2008 tokens (   10.14 ms per token)
  model_print_timings:        eval time =   110.76 ms /     2 runs   (   55.38 ms per token)
  model_print_timings:       total time = 21740.37 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 20368.36ms
  prediction   1, time: 56.26ms
  prediction   2, time: 54.49ms
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  67.054 ms
  perf_total_per_op_us[                     MUL] =  25.284 ms
  perf_total_per_op_us[                  REPEAT] = 292.125 ms
  perf_total_per_op_us[                    GELU] = 443.020 ms
  perf_total_per_op_us[                    NORM] =  26.823 ms
  perf_total_per_op_us[                 MUL_MAT] = 17845.975 ms
  perf_total_per_op_us[                     CPY] =   9.194 ms
  perf_total_per_op_us[                    VIEW] =   0.901 ms
  perf_total_per_op_us[                 PERMUTE] =   0.112 ms
  perf_total_per_op_us[                GET_ROWS] =   9.606 ms
  perf_total_per_op_us[                    ROPE] =  55.728 ms
  perf_total_per_op_us[              FLASH_ATTN] = 231.289 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  12.471 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.646 ms
  perf_total_per_op_us[                     MUL] =   0.246 ms
  perf_total_per_op_us[                    GELU] =  11.491 ms
  perf_total_per_op_us[                    NORM] =   1.230 ms
  perf_total_per_op_us[                     CPY] =   0.928 ms
  perf_total_per_op_us[                    VIEW] =   0.860 ms
  perf_total_per_op_us[                 PERMUTE] =   0.099 ms
  perf_total_per_op_us[                GET_ROWS] =   0.022 ms
  perf_total_per_op_us[                    ROPE] =   1.073 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.416 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.504 ms
  perf_total_per_op_us[           INNER PRODUCT] =  30.194 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.520 ms
  perf_total_per_op_us[                     MUL] =   0.172 ms
  perf_total_per_op_us[                    GELU] =  11.316 ms
  perf_total_per_op_us[                    NORM] =   1.207 ms
  perf_total_per_op_us[                     CPY] =   0.861 ms
  perf_total_per_op_us[                    VIEW] =   0.916 ms
  perf_total_per_op_us[                 PERMUTE] =   0.104 ms
  perf_total_per_op_us[                GET_ROWS] =   0.022 ms
  perf_total_per_op_us[                    ROPE] =   1.022 ms
  perf_total_per_op_us[              FLASH_ATTN] =   3.690 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.550 ms
  perf_total_per_op_us[           INNER PRODUCT] =  29.443 ms
  ========================================
  model_print_timings:        load time = 20171.18 ms
  model_print_timings:      sample time =     3.32 ms /     3 runs   (    1.11 ms per token)
  model_print_timings: prompt eval time = 19023.91 ms /  2008 tokens (    9.47 ms per token)
  model_print_timings:        eval time =   109.89 ms /     2 runs   (   54.95 ms per token)
  model_print_timings:       total time = 20349.47 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 19023.91ms
  prediction   1, time: 55.94ms
  prediction   2, time: 53.95ms
  ```
  </details>
- <details>
  <summary>
  <code>env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_falcon -m /home_local/dingyi/data/falcon-7b-pr410-q40.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-f16 -p "$LUOYU_PROMPT"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  80.116 ms
  perf_total_per_op_us[                     MUL] =  24.818 ms
  perf_total_per_op_us[                  REPEAT] = 303.144 ms
  perf_total_per_op_us[                    GELU] = 408.902 ms
  perf_total_per_op_us[                    NORM] =  23.654 ms
  perf_total_per_op_us[                 MUL_MAT] = 30113.444 ms
  perf_total_per_op_us[                   SCALE] = 3015.494 ms
  perf_total_per_op_us[                     CPY] =  53.191 ms
  perf_total_per_op_us[                    VIEW] =   0.860 ms
  perf_total_per_op_us[                 PERMUTE] =   0.471 ms
  perf_total_per_op_us[                GET_ROWS] =   9.801 ms
  perf_total_per_op_us[           DIAG_MASK_INF] = 182.767 ms
  perf_total_per_op_us[                SOFT_MAX] = 511.043 ms
  perf_total_per_op_us[                    ROPE] =  62.090 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.816 ms
  perf_total_per_op_us[                     MUL] =   0.273 ms
  perf_total_per_op_us[                    GELU] =  12.178 ms
  perf_total_per_op_us[                    NORM] =   1.747 ms
  perf_total_per_op_us[                 MUL_MAT] =  10.204 ms
  perf_total_per_op_us[                   SCALE] =   1.581 ms
  perf_total_per_op_us[                     CPY] =   5.653 ms
  perf_total_per_op_us[                    VIEW] =   0.813 ms
  perf_total_per_op_us[                 PERMUTE] =   0.442 ms
  perf_total_per_op_us[                GET_ROWS] =   0.028 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.122 ms
  perf_total_per_op_us[                SOFT_MAX] =   2.276 ms
  perf_total_per_op_us[                    ROPE] =   1.440 ms
  perf_total_per_op_us[           INNER PRODUCT] =  31.910 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.492 ms
  perf_total_per_op_us[                     MUL] =   0.180 ms
  perf_total_per_op_us[                    GELU] =  12.290 ms
  perf_total_per_op_us[                    NORM] =   1.795 ms
  perf_total_per_op_us[                 MUL_MAT] =  10.161 ms
  perf_total_per_op_us[                   SCALE] =   1.505 ms
  perf_total_per_op_us[                     CPY] =   5.593 ms
  perf_total_per_op_us[                    VIEW] =   0.767 ms
  perf_total_per_op_us[                 PERMUTE] =   0.422 ms
  perf_total_per_op_us[                GET_ROWS] =   0.029 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.138 ms
  perf_total_per_op_us[                SOFT_MAX] =   2.350 ms
  perf_total_per_op_us[                    ROPE] =   1.435 ms
  perf_total_per_op_us[           INNER PRODUCT] =  31.871 ms
  ========================================
  model_print_timings:        load time = 35948.11 ms
  model_print_timings:      sample time =     2.71 ms /     3 runs   (    0.90 ms per token)
  model_print_timings: prompt eval time = 34795.06 ms /  2008 tokens (   17.33 ms per token)
  model_print_timings:        eval time =   146.85 ms /     2 runs   (   73.43 ms per token)
  model_print_timings:       total time = 36152.77 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 34795.05ms
  prediction   1, time: 73.77ms
  prediction   2, time: 73.09ms
  ```
  </details>

In addition, it is tested that llama & gptj will preserve its behavior in this PR (q4j-int8-g128). 

## Dependency Change?
N/A

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/Ops.h [hunk 1]
```diff
@@ -85,6 +85,7 @@ enum ne_op {
   NE_OP_ALL_REDUCE,
   NE_OP_TP_CONCAT,
   NE_OP_DUMP_TENSOR,
+  NE_OP_DBG,
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/Ops.h:88]: DBG for ?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 2]
```diff
@@ -137,10 +140,12 @@ class ScaleExpAccSumFp32 {
     int ld_dst;  // #elements
     float scale;
     int causal_offset;  // offset for causal mask; negative value for disabling causal mask
+    float alibi_slope;  // m-factor in the alibi paper for current head: https://arxiv.org/abs/2108.12409
   };
 
   JBLAS_CODE forward(const float* src, const int src_step, const int M_offset, const int N_offset, const int M,
                      const int N, const Param& p) const {
+    assert(("alibi not supported!", p.alibi_slope == 0.f));
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:148]: "alibi" ?

**Review by airMeng — COMMENTED** [review]
where can I find explanation to these abbr?

**Comment by airMeng** [comment]
all model owners are aware of this?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:148]: This is for future integration of MPT.

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/Ops.h (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/Ops.h:88]: Debug

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/Ops.h:88]: OK, it's reasonable to add back the 2 letters for clarity.

**Comment by DDEle** [comment]
BTW, @a32543254 I added a flexible OP for debug experience. It allows developers to access forwarding result in the graph file (with help of non-capture lambda expression).

https://github.com/intel/intel-extension-for-transformers/commit/69ee8ebce4239646d4f1126ef2d9b3bfe3934d87#diff-fd662f3dbbde39aca9f4a1df0017d713ee7c5aa508025b36315c6defac152bc0R178-R184

**Review by airMeng — APPROVED** [review]

**Review by zhentaoyu — APPROVED** [review]

**Pull request merged**